networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/validation/AttestationValidator.java

```diff
@@ -0,0 +1,212 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.eth2.gossip.topics.validation;
+
+import static com.google.common.primitives.UnsignedLong.ONE;
+import static com.google.common.primitives.UnsignedLong.ZERO;
+import static tech.pegasys.artemis.datastructures.util.AttestationUtil.get_indexed_attestation;
+import static tech.pegasys.artemis.datastructures.util.AttestationUtil.is_valid_indexed_attestation;
+import static tech.pegasys.artemis.networking.eth2.gossip.topics.validation.ValidationResult.INVALID;
+import static tech.pegasys.artemis.networking.eth2.gossip.topics.validation.ValidationResult.SAVED_FOR_FUTURE;
+import static tech.pegasys.artemis.networking.eth2.gossip.topics.validation.ValidationResult.VALID;
+import static tech.pegasys.artemis.util.config.Constants.ATTESTATION_PROPAGATION_SLOT_RANGE;
+import static tech.pegasys.artemis.util.config.Constants.SECONDS_PER_SLOT;
+import static tech.pegasys.artemis.util.config.Constants.VALID_ATTESTATION_SET_SIZE;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.datastructures.operations.IndexedAttestation;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.util.CommitteeUtil;
+import tech.pegasys.artemis.storage.client.RecentChainData;
+import tech.pegasys.artemis.util.collections.ConcurrentLimitedSet;
+import tech.pegasys.artemis.util.collections.LimitStrategy;
+import tech.pegasys.artemis.util.config.Constants;
+
+public class AttestationValidator {
+
+  private static final UnsignedLong MILLIS_PER_SECOND = UnsignedLong.valueOf(10000);
+  private static final UnsignedLong MAXIMUM_GOSSIP_CLOCK_DISPARITY =
+      UnsignedLong.valueOf(Constants.MAXIMUM_GOSSIP_CLOCK_DISPARITY);
+
+  private final Set<ValidatorAndSlot> receivedValidAttestations =
+      ConcurrentLimitedSet.create(VALID_ATTESTATION_SET_SIZE, LimitStrategy.DROP_OLDEST_ELEMENT);
+  private final RecentChainData recentChainData;
+
+  public AttestationValidator(final RecentChainData recentChainData) {
+    this.recentChainData = recentChainData;
+  }
+
+  public ValidationResult validate(
+      final Attestation attestation, final UnsignedLong receivedOnSubnetId) {
+    ValidationResult validationResult = singleAttestationChecks(attestation, receivedOnSubnetId);
+    if (validationResult != VALID) {
+      return validationResult;
+    }
+
+    validationResult = singleOrAggregateAttestationChecks(attestation);
+    if (validationResult != VALID) {
+      return validationResult;
+    }
+
+    return firstValidAttestationCheck(attestation);
+  }
+
+  private ValidationResult firstValidAttestationCheck(final Attestation attestation) {
```

Comment:
nit: this might be a little bit more appropriate
```suggestion
  private ValidationResult addAndCheckFirstValidAttestation(final Attestation attestation) {
```

Reply:
Done.